test_rmw_implementation/test/test_graph_api.cpp

```diff
@@ -913,3 +916,103 @@ TEST_F(CLASSNAME(TestGraphAPI, RMW_IMPLEMENTATION), count_subscribers_with_bad_a
+  size_t count = 0u;
+  constexpr char service_name[] = "/test_service";
+
+  // A null node is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_clients(nullptr, service_name, &count));
+  rmw_reset_error();
+
+  // A node from a different implementation cannot be used to query.
+  const char * implementation_identifier = node->implementation_identifier;
+  node->implementation_identifier = "not-an-rmw-implementation-identifier";
+  EXPECT_EQ(RMW_RET_INCORRECT_RMW_IMPLEMENTATION, rmw_count_clients(node, service_name, &count));
+  node->implementation_identifier = implementation_identifier;
+  rmw_reset_error();
+
+  // A null topic name is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_clients(node, nullptr, &count));
+  rmw_reset_error();
+
+  // An invalid topic name is an invalid argument.
+  constexpr char invalid_service_name[] = "not a valid topic name !";
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_clients(node, invalid_service_name, &count));
+  rmw_reset_error();
+
+  // A null count is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_clients(node, service_name, nullptr));
+  rmw_reset_error();
+}
+
+TEST_F(CLASSNAME(TestGraphAPI, RMW_IMPLEMENTATION), count_services_with_bad_arguments) {
+  size_t count = 0u;
+  constexpr char service_name[] = "/test_service";
+  // A null node is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_services(nullptr, service_name, &count));
+  rmw_reset_error();
+
+  // A node from a different implementation cannot be used to query.
+  const char * implementation_identifier = node->implementation_identifier;
+  node->implementation_identifier = "not-an-rmw-implementation-identifier";
+  EXPECT_EQ(RMW_RET_INCORRECT_RMW_IMPLEMENTATION, rmw_count_services(node, service_name, &count));
+  node->implementation_identifier = implementation_identifier;
+  rmw_reset_error();
+
+  // A null topic name is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_services(node, nullptr, &count));
+  rmw_reset_error();
+
+  // An invalid topic name is an invalid argument.
+  constexpr char invalid_service_name[] = "not a valid topic name !";
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_services(node, invalid_service_name, &count));
+  rmw_reset_error();
+
+  // A null count is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_services(node, service_name, nullptr));
+  rmw_reset_error();
+}
+
+TEST_F(CLASSNAME(TestGraphAPI, RMW_IMPLEMENTATION), count_clients_and_services) {
+  constexpr char service_name[] = "/test_service";
+  const rosidl_service_type_support_t * ts =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, BasicTypes);
+  size_t count = 0u;
+
+  rmw_service_t * srv = rmw_create_service(
+    other_node, ts, service_name, &rmw_qos_profile_services_default);
+  ASSERT_NE(nullptr, srv) << rmw_get_error_string().str;
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
I'd really prefer not to have hard-coded sleep times in the tests.  Over time we've found that they are both too long and too short.

They are too long because if the action completes very quickly, we are wasting time.  They are too short because of timings within individual machines, and in CI, which leads to them being flaky.

Elsewhere, what we do is a tight loop where we check for the condition.  If the condition is satisfied, we break out of the loop.  If it is not satisfied, we sleep for a very short amount of time (1 or 10 milliseconds), then check again.  And then we have an overall timeout (2 or 5 seconds), just to ensure that we don't run forever in the case of a bug.  I'd much prefer if we did something closer to that here.

Reply:
I really appreciate your cool explanation! fixed